Aikido.Zen.Core/Helpers/RateLimitingHelper.cs

```diff
@@ -27,36 +32,115 @@ public static bool IsAllowed(string key, int windowSizeInMS, int maxRequests)
+                timestamps.RemoveAll(timestamp => currentTime - timestamp > windowSizeInMS);
 
-                if (elapsedTime >= windowSizeInMS)
+                // Add current timestamp
+                timestamps.Add(currentTime);
+
+                // Update the cache with filtered timestamps
+                RateLimitedItems.Set(key, timestamps);
+
+                // Check if the number of requests is within limits
+                return timestamps.Count <= maxRequests;
+            }
+        }
+
+        /// <summary>
+        /// Checks if a route matches a wildcard pattern
+        /// </summary>
+        /// <param name="routePattern">The route pattern that may contain wildcards (*)</param>
+        /// <param name="actualRoute">The actual route to check against the pattern</param>
+        /// <returns>True if the route matches the pattern, false otherwise</returns>
+        public static bool IsWildcardMatch(string routePattern, string actualRoute)
+        {
+            if (string.IsNullOrEmpty(routePattern)) return false;
+            if (string.IsNullOrEmpty(actualRoute)) return false;
+
+            // Convert the wildcard pattern to a regex pattern
+            string regexPattern = "^" + Regex.Escape(routePattern).Replace("\\*", ".*") + "$";
+            return Regex.IsMatch(actualRoute, regexPattern, RegexOptions.IgnoreCase);
+        }
+
+        /// <summary>
+        /// Determines if a request should be allowed based on all applicable rate limiting rules
+        /// </summary>
+        /// <param name="context">The context of the request</param>
+        /// <param name="endpoints">The filtered list of endpoints</param>
+        /// <returns>A tuple containing: (isAllowed, effectiveConfig) where effectiveConfig is the config that caused rate limiting (if any)</returns>
+        public static (bool isAllowed, RateLimitingConfig effectiveConfig) IsRequestAllowed(
+            Context context,
+            IEnumerable<EndpointConfig> endpoints)
+        {
+            if (string.IsNullOrEmpty(context?.Method) || string.IsNullOrEmpty(context?.Route))
+            {
+                return (true, null);
+            }
+
+            if (endpoints == null || !endpoints.Any())
+            {
+                return (true, null);
+            }
+
+            // Get the user ID or IP address for the key
+            string userOrIp = context.User?.Id ?? context.RemoteAddress ?? "unknown";
+
+            // Find endpoints that have rate limiting enabled
+            var rateLimitedEndpoints = endpoints
+                .Where(e => e.RateLimiting != null && e.RateLimiting.Enabled)
+                .ToList();
+
+            if (!rateLimitedEndpoints.Any())
+            {
+                return (true, null);
+            }
+
+            // Find exact match first
+            var exactMatch = rateLimitedEndpoints.FirstOrDefault(e => e.Method == context.Method && e.Route == context.Route);
```

Comment:
this stuff as well should be in a unified helper class, the rate limiting sort not